Add maintenance socket

margdoc opened the pull request:
It enables interaction with the node through CQL protocol without authentication. It gives full-permission access.
The maintenance socket is available by Unix domain socket with file permissions `755`, thus it is not accessible from outside of the node and from other POSIX groups on the node.
It is created before the node joins the cluster.

To set up the maintenance socket, use the `maintenance-socket` option when starting the node.

* If set to `ignore` maintenance socket will not be created.
* If set to `workdir` maintenance socket will be created in `<node's workdir>/cql.maintenance`.
* Otherwise maintenance socket will be created in the specified path.

The default value is `workdir`.

## Connect to maintenance socket

* With python driver

```python
from cassandra.cluster import Cluster
from cassandra.connection import UnixSocketEndPoint

cluster = Cluster([UnixSocketEndPoint("<node's workdir>/cql.m")])
session = cluster.connect()
```

denesb commented on db/config.cc:
```diff
@@ -747,6 +747,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , native_transport_port(this, "native_transport_port", "cql_port", value_status::Used, 9042,
         "Port on which the CQL native transport listens for clients.")
+    , maintenance_socket(this, "maintenance_socket", value_status::Used, "/tmp/scylla/cql.maintenance",
+        "The Unix Domain Socket a node uses for maintenance port. If set to 'ignore', the node will not start the maintenance port")
```
Should have a brief mention about this being a full-permission thing.

denesb replied:
I think the default should be that it is disabled. And we should maybe make this live-update and add a REST API to enable/disable it.

kbr-scylla replied:
> I think the default should be that it is disabled. 

Why? IIRC @avikivity had the opposite idea.

denesb replied:
Ok, I don't have a strong opinion.

denesb commented on configure.py:
```diff
@@ -1130,6 +1130,7 @@ def find_ninja():
                 'auth/permissions_cache.cc',
                 'auth/service.cc',
                 'auth/standard_role_manager.cc',
+                'auth/maintenance_port_role_manager.cc',
```
The above two should go into the previous commit, no?

denesb commented on maintenance_mode.hh:
```diff
@@ -0,0 +1,15 @@
+
+using namespace seastar;
+
+using maintenance_port_enabled = bool_class<class maintenance_port_enabled_tag>;
```
Can't this go to `db/config.hh`? I guess I'll see later how it is used.

denesb commented on transport/controller.cc:
```diff
@@ -112,47 +114,61 @@ future<> controller::do_start_server() {
-            utils::configure_tls_creds_builder(*cred, std::move(ceo)).get();
-
-            logger.info("Enabling encrypted CQL connections between client and server");
+        if (!_maintenance_port) {
```
I'm confused at all these changes. I expected this maintenance port to be a pure addition, but it looks like it is exclusive with the regular CQL port? I don't think that makes sense. The regular CQL port can already be disabled via config and nodetool, the maintenance socket should be just another optional way for CQL traffic to be received.

BTW, maintenance_port is called maintenance_socket elsewhere, please use consistent naming.

kbr-scylla replied:
If I understand correctly, there will be two instances of `cql_server_controller`, one corresponding to regular port, one to maintenance port.

The maintenance port one has a bunch of stuff done differently / disabled (like authentication).

> but it looks like it is exclusive with the regular CQL port? 

Why? Doesn't seem like it to me

denesb replied:
Ah, it is really hard to see what is going an due to all the changes in this method.

kbr-scylla replied:
There are not many actual changes, most of the diff comes from changing the indentation, IIUC. @margdoc please split the indentation change to a separate commit

denesb commented on auth/service.cc:
```diff
@@ -189,9 +194,13 @@ future<> service::create_keyspace_if_missing(::service::migration_manager& mm) c
+        ? make_ready_future<>()
+        : once_among_shards([this, &mm] {
+            return create_keyspace_if_missing(mm);
+        });
```
I don't understand why we need to disable this when the maintenance port is enabled. This seems like a layering violation to me.

kbr-scylla replied:
From what I understand: all queries go through cql controller which underneath has a reference to auth service.

We want auth to be disabled for maintenance port.

So the maintenance-port instance of cql controller has a reference to a "dummy" auth service that does not create a keyspace and so on. (It cannot, because at the moment of opening maintenance port, we don't have access to the cluster; and the maintenance port is supposed to also be used in situations where we won't have access to the cluster later, so we cannot do auth over distributed tables).

The "old" auth service will still do all of this stuff.

kbr-scylla replied:
@margdoc I think what's confusing is the name `maintenance_port_enabled`.

It's not intuitive that there will be two instances of auth service where one has `maintenance_port_enabled = false` and another has `maintenance_port_enabled = true`. The port cannot be both enabled and disabled at the same time.

The name should instead somehow signify that this instance of the service is "associated to" the maintenance port. Maybe `used_by_maintenance_port`?

margdoc replied:
It is disabled only for maintenance_auth_service which doesn't need that and it can't be created before joining group0.
When regular auth_service is started, it is called as before.

denesb replied:
Maybe we should invent a better name for this "maintenance_port" parameter.

denesb commented on main.cc:
```diff
@@ -12,6 +12,10 @@
 
```
missing: motivation and explanation of why this is fine.

denesb reviewed (changes requested):
It feels like this PR is doing too much. It doesn't just add yet another way of connecting to a node via CQL, it shuffles the init order and turns things on/off based on whether the maintenance socket was enabled or not. I think all these kind of changes, should come later, as part of the maintenance mode PR. This PR should just add the maintenance socket as a pure addition, without changing any of the existing mechanisms of the node.

denesb commented:
BTW, I really like the basic idea. Unix Socket seems like the perfect candidate for a special CQL channel.

kbr-scylla commented on test/pylib/scylla_cluster.py:
```diff
@@ -694,6 +694,8 @@ async def add_server(self, replace_cfg: Optional[ReplaceConfig] = None,
 
         extra_config: dict[str, Any] = config.copy() if config else {}
+        if "maintenance_socket" not in extra_config:
+            extra_config["maintenance_socket"] = "ignore"
```
>  test.py: disable maintenance port by default 

Please write motivation in the commit message

margdoc replied:
Removed this commit, since it is not needed to set this value as ignore by default in tests.

kbr-scylla commented on transport/controller.cc:
```diff
@@ -112,47 +114,61 @@ future<> controller::do_start_server() {
         std::vector<listen_cfg> configs;
         int native_port_idx = -1, native_shard_aware_port_idx = -1;
 
-        if (cfg.native_transport_port.is_set() ||
```
@margdoc This will be much easier to review if you split this commit as follows:
- the first commit adds the `if (!_maintenance_port)` but does not change indentation of code inside the `if` (from what I see, you're mostly putting a branch around pre-existing code)
- the second commit fixes the indentation inside the branch

denesb replied:
BTW:
"If a method has a boolean parameter and its content is enclosed in a big if on that parameter, it should really be two methods"
Julius Caesar, 40 BC

kbr-scylla commented on transport/controller.cc:
```diff
@@ -112,47 +114,61 @@ future<> controller::do_start_server() {
+            }
+        } else {
+            auto socket = cfg.maintenance_socket();
+            auto r = ::unlink(socket.c_str());
```
Please add a comment why this unlinking is necessary.

kbr-scylla commented on auth/maintenance_port_role_manager.cc:
```diff
@@ -0,0 +1,113 @@
+
+namespace auth {
+
+constexpr std::string_view maintenance_port_role_manager_name = "com.scylladb.auth.MaintenanceModeRoleManager";
```
"MaintenanceMode" appeared here. Should be `MaintenancePortRoleManager`. Or `MaintenanceSocketRoleManager` if we go with "socket" instead of "port" and other places are renamed too.

kbr-scylla commented on auth/maintenance_port_role_manager.hh:
```diff
@@ -0,0 +1,70 @@
+
+extern const std::string_view maintenance_port_role_manager_name;
+
+class maintenance_port_role_manager final : public role_manager {
```
Please describe the purpose of this in a comment.

kbr-scylla commented on auth/maintenance_port_role_manager.cc:
```diff
@@ -0,0 +1,113 @@
+
+future<> maintenance_port_role_manager::create(std::string_view role_name, const role_config&) {
+        return make_exception_future<>(
+        std::runtime_error("CREATE operation is not supported by MaintenanceModeRoleManager"));
```
"Mode" word appearing everywhere.

kbr-scylla replied:
Also, perhaps we should make the error messages more high-level / user-friendly? "RoleManager" Such as "role manager: CREATE operation not supported through maintenance port" (/"socket").

kbr-scylla commented on auth/service.cc:
```diff
@@ -30,6 +30,7 @@
 #include "db/functions/function_name.hh"
 #include "exceptions/exceptions.hh"
 #include "log.hh"
+#include "maintenance_mode.hh"
```
>  auth/service: don't create system_auth keyspace when creating maintenance port

More context should be provided by the commit message: the fact that there will be two auth services, and only the new auth service will not create system_auth.

kbr-scylla commented on main.cc:
```diff
@@ -1639,6 +1644,67 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            cql_transport::controller cql_server_ctl(auth_service, mm_notifier, gossiper, qp, service_memory_limiter, sl_controller, lifecycle_notifier, *cfg, scheduling_group_key_create(cql_sg_stats_cfg).get0(), maintenance_port_enabled::no);
+            cql_transport::controller cql_maintenance_server_ctl(maintenance_auth_service, mm_notifier, gossiper, qp, service_memory_limiter, sl_controller, lifecycle_notifier, *cfg, scheduling_group_key_create(maintenance_cql_sg_stats_cfg).get0(), maintenance_port_enabled::yes);
+
+            auto start_cql = [&dbcfg] (cql_transport::controller& controller, std::any& stop_cql) {
```
Code movements and refactoring (such as extraction of `start_cql` function) should be in a separate commit than the new stuff.

kbr-scylla commented on test/auth_cluster/test_maintenance_port.py:
```diff
@@ -0,0 +1,62 @@
+    try:
+        cluster = Cluster([server.ip_addr])
+        session = cluster.connect()
+        pytest.fail("Should have failed")
```
More pythonic would be to put this into `else` branch.
The message should be more descriptive, "Should have failed" says nothing more than using `pytest.fail("")`.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - Build
:x: - Unit Tests
:x: - Sanity Tests
#### Failed Tests (254/13039):
- [test_maintenance_port](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_maintenance_port/Tests___Unit_Tests___test_maintenance_port) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_port)
- [test_maintenance_port](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_maintenance_port/Tests___Unit_Tests___test_maintenance_port_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_port)
- [test_maintenance_port](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_maintenance_port/Tests___Unit_Tests___test_maintenance_port_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_port)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_misconfigured_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_misconfigured_storage) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_misconfigured_storage)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_misconfigured_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_misconfigured_storage_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_misconfigured_storage)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4887/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4887/
- Duration: 5 hr 0 min
- Builder: spider1.cloudius-systems.com

margdoc commented:
v2:
* maintenance-socket flag can be set to "workdir" and then the maintenance socket will be created in node's work directory (it fixes other tests which failed due to sharing the same maintenance socket). Make this value as default.
* remove setting maintenance-socket flag to "ignore" in tests
* marked test_maintenance_socket with xfail (python driver regression)
* split commit changing cql_controller to two seperate commits, where the second one only changes indentation
* split commit changing main.cc
* s/maintanance_port/maintenance_socket
* maintenance_socket_role_manager: make error messages more user-friendly
* maintenance_socket_role_manager: add comment about this class
* cql controller, auth service: s/maintenance_socket_enabled/used_by_maintenance_socket
* add more context about changes made to cql controller and auth service
*  moved maintanance_socket_enabled to db/config
* add comment about ::unlink
* test: move pytest.fail statements to "else" branches and make it more descriptive

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/16172)

**Changed Files:**
- [operating-scylla/_common/tools_index](http://previews.docs.scylladb.com/scylladb/16172/operating-scylla/_common/tools_index)
- [operating-scylla/admin-tools/index.html](http://previews.docs.scylladb.com/scylladb/16172/operating-scylla/admin-tools/index.html)
- [operating-scylla/admin-tools/maintenance-socket](http://previews.docs.scylladb.com/scylladb/16172/operating-scylla/admin-tools/maintenance-socket)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5138/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5138/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6517/)
#### Failed Tests (198/13074):
- [test_maintenance_port](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_maintenance_port/Tests___Unit_Tests___test_maintenance_port) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_port)
- [test_maintenance_port](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_maintenance_port/Tests___Unit_Tests___test_maintenance_port_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_port)
- [test_maintenance_port](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_maintenance_port/Tests___Unit_Tests___test_maintenance_port_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_maintenance_port)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_misconfigured_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_misconfigured_storage) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_misconfigured_storage)
- [test_memtable_flush_retries](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_memtable_flush_retries) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_memtable_flush_retries)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_misconfigured_storage](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5138/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_misconfigured_storage_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_misconfigured_storage)
#### Build Details:
- Duration: 5 hr 4 min
- Builder: spider2.cloudius-systems.com

kbr-scylla commented:
Weird, the last CI build ran on commit 227901f9a769857526d38d93b21a3055c08452c8 which used 5d79f52e1b2279dc13f504ba708f5b480532fecb -- that's an old version of this PR, before the last two force pushes

The last commit in the newest version of this PR is 404ece28c25520bad6eb7b3251734bf52661d9dd

@benipeled any idea why CI is not using the newest version?

kbr-scylla commented:
I started another rebuild: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5152/

kbr-scylla commented:
Ugh, the Jenkins console output in that last CI build is huge, >200MB. It may have inlined all the test.py failures...

kbr-scylla commented:
```
[2023-12-08T20:43:35.005Z] Checking out Revision 227901f9a769857526d38d93b21a3055c08452c8 (refs/remotes/origin/pr/16172/merge)
[2023-12-08T20:43:35.467Z] Commit message: "Merge 5d79f52e1b2279dc13f504ba708f5b480532fecb into 4505a86f46dfafa039246369c46140a714efadbd"
```

kbr-scylla commented:
The new job is also taking the old commit, aborted.
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5152/consoleText
```
[2023-12-11T08:58:59.427Z] Checking out Revision 227901f9a769857526d38d93b21a3055c08452c8 (refs/remotes/origin/pr/16172/merge)
[2023-12-11T08:58:59.891Z] Commit message: "Merge 5d79f52e1b2279dc13f504ba708f5b480532fecb into 4505a86f46dfafa039246369c46140a714efadbd"
```

is it because there's a conflict in main?

margdoc commented:
v3:
* rebased on master

kbr-scylla commented:
Newest build uses the correct commit:
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5153/console
```
10:11:36  Checking out Revision f5bb3649ed5623f1880dc5d6baee82f14e143664 (refs/remotes/origin/pr/16172/merge)
10:11:37  Commit message: "Merge 206ffaa1baaa9c40ef1b2908ebf7044edee9ee60 into f30b5473abf54987cb016134fec74342fe10e8d1"
```

margdoc commented:
Found a problem testing it locally.
```c
/* Structure describing the address of an AF_LOCAL (aka AF_UNIX) socket.  */
struct sockaddr_un
  {
    __SOCKADDR_COMMON (sun_);
    char sun_path[108];		/* Path name.  */
  };
```
Path of Unix domain socket can have at most 108 bytes. When a path provided by the flag `maintenance-socket` is too long (for example, work directory of the node is too long) it is trimmed to fit in the `sockaddr_un` structure. It can cause problem, because unlink is called on original path, but bind on the trimmed path. When the node is restarted, unlink will return error that the file doesn't exist (this is not an error for us, because usually this file won't exist) and bind will hang, because previous uds still exists.

Decided to check if path is too long and if it is, print an error and exit so that user can set it manually to somewhere else.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5153/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5153/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6543/)
#### Failed Tests (197/20185):
- [test_truncate_after_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/cql_tests/TestTruncate/Tests___dtest___test_truncate_after_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_after_restart)
- [test_simple_single_node_write_read[SMP=1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/simple_no_driver_test/TestSimple/Tests___dtest___test_simple_single_node_write_read_SMP_1_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_single_node_write_read[SMP=1])
- [test_simple_kill_new_node_while_bootstrapping](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_kill_new_node_while_bootstrapping) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_kill_new_node_while_bootstrapping)
- [test_simple_removenode_3[api-False]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_3_api_False_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_3[api-False])
- [test_tracing_shutdown](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/cql_tracing_test/TestCqlTracing/Tests___dtest___test_tracing_shutdown) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tracing_shutdown)
- [test_remove_garbage_members_from_group0_after_abort_decommission[Announcing_that_I_have_left_the_ring-]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_remove_garbage_members_from_group0_after_abort_decommission_Announcing_that_I_have_left_the_ring__) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_members_from_group0_after_abort_decommission[Announcing_that_I_have_left_the_ring-])
- [test_simple_removenode_3[api-True]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_3_api_True_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_3[api-True])
- [test_simple_removenode_3[nodetool-False]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_3_nodetool_False_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_3[nodetool-False])
- [test_tracing_startup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/cql_tracing_test/TestCqlTracing/Tests___dtest___test_tracing_startup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tracing_startup)
- [test_remove_garbage_members_from_group0_after_abort_decommission[became_a_group_0_non-voter-]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5153/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_remove_garbage_members_from_group0_after_abort_decommission_became_a_group_0_non_voter__) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_members_from_group0_after_abort_decommission[became_a_group_0_non-voter-])
#### Build Details:
- Duration: 4 hr 52 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
Looks like the `workdir` option for maintenance_socket path is incompatible with existing configurations; it assumes that the config also provides `workdir` in order to create the socket in that workdir. But old configurations may not provide `workdir`/`workdir,W`, instead providing separate directories for data, commitlog and so on.

Therefore I suggest making `ignore` the default in db/config.cc. But in test/pylib/scylla_cluster.py, pass `workdir` for maintenance_socket path. And also modify conf/scylla.yaml to provide `workdir`. This way:
- we will be opening maintenance_socket in all topology tests by default
- in new clusters too
- in old clusters with old configurations, we will not be opening it. Admin will have to modify their conf/scylla.yaml and chose a path, whether it is workdir or something else
- in dtests we will not be opening it for now; we can send a PR to scylla-dtest to use workdir instead of data_dir etc., and then also open maintenance socket

I think it's a good compromise between testing the feature and not breaking existing configurations / production clusters.

kostja commented:
I like the suggestion. Please do not ignore dtest, or it will backfire. File a ticket to similarly update dtest and @temichus team can handle it.

temichus commented:
> I like the suggestion. Please do not ignore dtest, or it will backfire. File a ticket to similarly update dtest and @temichus team can handle it.

as i see in https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6543/testReport/ all tests failed at the start_scylla with exactly same  error that log message at start was not found:
```
ccmlib.node.TimeoutError: 11 Dec 2023 10:09:31 [node1] Missing: ['Starting listening for CQL clients|storage_service .* Starting to bootstrap|(compaction|database) -.*Resharding']:
Scylla version 5.5.0~dev-0.20231211.f5bb3649ed56 w.....
```

@margdoc, Is this message changed in this PR

kbr-scylla commented:
@temichus the reason is that the nodes are hanging trying to open the unix domain socket, because they are all trying to use the same path for the socket (inside `workdir` which in dtest is the default /var/lib/scylla). It's a problem with this PR that didn't take into account the configuration that dtests are using.

denesb commented:
> Looks like the `workdir` option for maintenance_socket path is incompatible with existing configurations; it assumes that the config also provides `workdir` in order to create the socket in that workdir. But old configurations may not provide `workdir`/`workdir,W`, instead providing separate directories for data, commitlog and so on.

I'm fine with this, but please make sure that:
* This dependency of `maintenance_socket`'s `workdir` option on the `workdir` config item is documented in the config item description
* When `maintenance_socket=workdir` and `workdir` is not set, make sure we print a warning stating that the socket cannot be created because `workdir` is not set.

kbr-scylla commented:
> When maintenance_socket=workdir and workdir is not set, make sure we print a warning stating that the socket cannot be created because workdir is not set.


Unfortunately `workdir,W` has a default value:
```cpp
    , work_directory(this, "workdir,W", value_status::Used, "/var/lib/scylla",
        "The directory in which Scylla will put all its subdirectories. The location of individual subdirs can be overridden by the respective *_directory options.")
```

so it is pretty much always set. The warning will never fire.

The problem in dtests was that all nodes were pointing to the same default workdir, so they tried to listen on the same UDS.

Documenting the dependency of course should be done.

denesb commented:
> > When maintenance_socket=workdir and workdir is not set, make sure we print a warning stating that the socket cannot be created because workdir is not set.
> 
> Unfortunately `workdir,W` has a default value:
> 
> ```c++
>     , work_directory(this, "workdir,W", value_status::Used, "/var/lib/scylla",
>         "The directory in which Scylla will put all its subdirectories. The location of individual subdirs can be overridden by the respective *_directory options.")
> ```
> 
> so it is pretty much always set. The warning will never fire.
> 
> The problem in dtests was that all nodes were pointing to the same default workdir, so they tried to listen on the same UDS.
> 
> Documenting the dependency of course should be done.

Can we check that workdir is a prefix of datadir and make a decision based on that?

denesb commented:
> Can we check that workdir is a prefix of datadir and make a decision based on that?

Actually, I think defaulting to `ignore` is the better option. Heuristics will just cause more confusion.

kbr-scylla commented:
> Actually, I think defaulting to ignore is the better option. Heuristics will just cause more confusion.

Having heuristics for better UX are compatible with defaulting to ignore.

> Can we check that workdir is a prefix of datadir and make a decision based on that?

You mean `workdir,W` is a prefix of `data_file_directories`? I don't think we could introduce such a check, it would break existing configurations. Besides I'm not sure how it could help for `maintenance_socket.

denesb commented:
> > Actually, I think defaulting to ignore is the better option. Heuristics will just cause more confusion.
> 
> Having heuristics for better UX are compatible with defaulting to ignore.
> 
> > Can we check that workdir is a prefix of datadir and make a decision based on that?
> 
> You mean `workdir,W` is a prefix of `data_file_directories`? I don't think we could introduce such a check, it would break existing configurations. Besides I'm not sure how it could help for `maintenance_socket.

I don't mean a general check. This would detect the case where `workdir` was just left in its default value and only `datadir`  was set. Or maybe the config item has a `defaulted()` method, like `bpo` has, which returns `true` if the option was left with its default value, and not set by the user.

margdoc commented:
v4:
* db/config: maintenance-socket set to ignore by default
* conf/scylla.yaml: maintenance-socket set to workdir by default
* updated docs
* test: maintenance-socket set to workdir by default
* throw exception if socket path is too long or something exists in this path

margdoc commented:
With Kamil we found out that dtest takes `maintenance-socket` option from scylla.yaml. It was set to `workdir`, thus each node in dtest had socket path set to `/var/lib/scylla/cql.m` which is the problem we want to fix.
We decided to set it to ignore in scylla.yaml for now. When it is fixed in dtest (`maintenance-socket` is set to some unique value or `workdir` flag is set), `maintenance-socket` in scylla.yaml will be changed to `workdir`.

v5:
* conf/scylla.yaml: maintenance-socket set to ignore by default

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5269/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6672/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5269/)
#### Build Details:
- Duration: 2 hr 24 min
- Builder: spider1.cloudius-systems.com

kbr-scylla commented on auth/service.cc:
```diff
@@ -131,7 +131,8 @@ service::service(
         ::service::migration_notifier& mn,
```
Re ` auth/service: don't create system_auth keyspace when used by maintenance socket` commit message

> The maintenance socket is created before joining the cluster. When an auth service
is started it creates system_auth keyspace if it's missing. It is not synchronized
with other nodes, because this node hasn't joined the group0 yet. Thus a node has
a mismatched schema and is unable to join the cluster.

I don't understand this description -- it's written as if it was describing current state of Scylla. But obviously nodes are able to join clusters and system_auth is created using group 0?

I think some context is missing.

kbr-scylla replied:
It would be a bit better if before mentioning "maintenance auth service" you introduced the concept. The first paragraph assumes that the reader knows what "maintenance auth service" is, but that concept is only introduced in the last paragraph.

The best order would be:
- explain that we want to create a separate auth service for maintenance socket (and why)
- then explain that there's a problem with that -- this happens too early and therefore blah blah
- finally the solution

But it's minor, the current commit message is good enough.

kbr-scylla commented on main.cc:
```diff
@@ -1643,6 +1643,32 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            perm_cache_config.refresh = std::chrono::milliseconds(cfg->permissions_update_interval_in_ms());
+
+            auto start_auth_service = [&mm] (sharded<auth::service>& auth_service, std::any& stop_auth_service) {
+                supervisor::notify("starting auth service");
```
Won't this mean we'll have two "starting auth service" messages, when you also use it later for the second "dummy" auth service?

I think we should include some context into the message so we distinguish the step of starting regular auth service from the maintenance-socket auth service.

Same for the `defer_verbose_shutdown` call below, the string should probably be different depending on the context.

kbr-scylla commented on main.cc:
```diff
@@ -1643,6 +1643,32 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            };
+
+            auto start_cql = [&dbcfg] (cql_transport::controller& controller, std::any& stop_cql) {
+                supervisor::notify("starting native transport");
```
Same comment regarding start and stop messages as above in start_auth_service.

kbr-scylla commented on main.cc:
```diff
@@ -1669,6 +1673,25 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            maintenance_auth_service.start(perm_cache_config, std::ref(qp), std::ref(mm_notifier), std::ref(mm), maintenance_auth_config, db::maintenance_socket_enabled::yes).get();
+
+            scheduling_group_key_config maintenance_cql_sg_stats_cfg =
+            make_scheduling_group_key_config<cql_transport::cql_sg_stats>(db::maintenance_socket_enabled::yes);
```
Can someone fluent with the language of scheduling groups review this part?
@denesb?

kbr-scylla replied:
or @avikivity

avikivity replied:
The second line in a multi-line statement needs a double indent.

But, why do we need separate statistics? The maintenance port will usually be quiet. We can just allow them to share statistics. Pushing out statistics comes with a large cost, and these are already split over service levels.

kbr-scylla replied:
Ok I understand this now after reading seastar test for scheduling group specific values. (The comments somehow didn't help me)

`scheduling_group_key_create` creates a map: the key is the scheduling group, the value is of type specified by `make_scheduling_group_key_config`.

So we have a value of given type in each scheduling group, and we can access the value of a given scheduling group using `scheduling_group_get_specific`.

So this will create a new stats object per scheduling group. This per-sg stats object will be then used by the new controller.

It's true that we create a new stats object, but these stats are write-only -- we don't register metrics based on them. The stats object is just there to avoid `if`fing every place we update stats in controller code (if stats object present / not used by maintenance socket, update stats...)

kbr-scylla commented on test/auth_cluster/test_maintenance_socket.py:
```diff
@@ -0,0 +1,66 @@
+
+    server = await manager.server_add(config=config)
+    workdir = await manager.server_get_workdir(server.server_id)
+    socket = workdir + "/cql.maintenance"
```
Looks like the test is outdated now.
But I think we can leave fixing it until after the Python driver issue is fixed and new version of the driver is released.

kbr-scylla commented on docs/operating-scylla/admin-tools/maintenance-socket.rst:
```diff
@@ -0,0 +1,31 @@
+    from cassandra.cluster import Cluster
+    from cassandra.connection import UnixSocketEndPoint
+    
+    cluster = Cluster([UnixSocketEndPoint("<node's workdir>/cql.m")])
```
You should probably combine this with usage of `WhiteListRoundRobinPolicy` to prevent the driver from connecting to / querying other nodes in the cluster.

margdoc replied:
Changed to `HostFilterPolicy`, since `WhiteListRoundRobinPolicy` doesn't accept unix domain socket path

kbr-scylla commented:
@avikivity @denesb please (re-)review

margdoc commented:
v6:
* updated comments and commit messages
* updated documentation with load balancing policy
* updated logging in main.cc
* transport/controller.cc: listen_cfg: designated field initialization, native_port_idx variable only for regular cql port
* transport/server: s/_used_by_maintenance_socket/used_by_maintenance_socket (only constructors arguments)
* fixed maintenance socket path in the test

avikivity commented:
> It enables interaction with the node through CQL protocol without authentication. It gives full-permission access. The maintenance socket is available by Unix domain socket with file permissions `755`, thus it is not accessible from outside of the node and from other POSIX groups on the node. It is created before the node joins the cluster.

This will have to change; I'm okay with a follow-up. Permissions should be 660 to allow a scyllaadm's group to connect. I don't think the execute permission is required.


> 
> To set up the maintenance socket, use the `maintenance-socket` option when starting the node.
> 
> * If set to `ignore` maintenance socket will not be created.
> * If set to `workdir` maintenance socket will be created in `<node's workdir>/cql.maintenance`.
> * Otherwise maintenance socket will be created in the specified path.
> 
> The default value is `workdir`.
> 
> ## Connect to maintenance socket
> * With python driver
> 
> ```python
> from cassandra.cluster import Cluster
> from cassandra.connection import UnixSocketEndPoint
> 
> cluster = Cluster([UnixSocketEndPoint("<node's workdir>/cql.m")])
> session = cluster.connect()
> ```

Please add a cqlsh example.

avikivity commented on conf/scylla.yaml:
```diff
@@ -591,6 +591,14 @@ consistent_cluster_management: true
+#            where <scylla's workdir> is a path defined by the workdir configuration option,
+# * <socket path>: the node will open the maintenance socket on the path <socket path>.
+maintenance_socket: ignore
+
```
Why not `workdir`?

avikivity replied:
And no need to include it in scylla.yaml, can just inherit the default from configuration.

avikivity replied:
from .cc

kbr-scylla replied:
It breaks dtests unfortunately, because all clusters in a dtest run are using the same workdir (ccm when creating Scylla configuration, instead of setting `workdir`, they are setting data, commitlog etc directories separately.

I propose we fix this in a follow-up -- we'll need a change in dtest/ccm so each cluster creates its own workdir, and then we can switch this to workdir.

avikivity commented on auth/maintenance_socket_role_manager.cc:
```diff
@@ -0,0 +1,108 @@
+future<> maintenance_socket_role_manager::grant(std::string_view grantee_name, std::string_view role_name) {
+    return operation_not_supported_exception("GRANT");
+}
+
```
We might decide to allow authentication after all (a useful way to fix broken configuration), but let's leave it like that for now.

avikivity commented on docs/operating-scylla/admin-tools/maintenance-socket.rst:
```diff
@@ -0,0 +1,35 @@
+
+Connect to maintenance socket
+-----------------------------
+
```
Need cqlsh example, much more likely to be used with the maintenance socket.

kostja replied:
cqlsh doesn't work afaics

avikivity commented:
Looks very good. But we can drop the split stats, try `curl localhost:9180/metrics | wc ` from a very large active node and you'll see why.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5304/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6709/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5304/)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented on transport/server.cc:
```diff
@@ -188,9 +188,13 @@ event::event_type parse_event_type(const sstring& value)
+cql_sg_stats::cql_sg_stats(db::maintenance_socket_enabled used_by_maintenance_socket)
     : _cql_requests_stats(static_cast<uint8_t>(cql_binary_opcode::OPCODES_COUNT))
 {
+    if (used_by_maintenance_socket) {
```
@avikivity looks like metrics are disabled on the maintenance socket. I think it makes sense, having combined metrics would most likely require a larger refactor, but I don't think there is a point in having metrics on maintenance socket -- one shouldn't run any serious workloads on it.

kbr-scylla commented on docs/operating-scylla/admin-tools/maintenance-socket.rst:
```diff
@@ -0,0 +1,35 @@
+    socket = "<node's workdir>/cql.m"
+    cluster = Cluster([UnixSocketEndPoint(socket)],
+                      # Driver tries to connect to other nodes in the cluster, so we need to filter them out.
+                      load_balancing_policy=HostFilterPolicy(RoundRobinPolicy(), lambda h: h.address == socket))
```
IIRC cqlsh didn't work, but I don't remember what was the reason?

Maybe it's because it uses `WhiteListRoundRobinPolicy` underneath?

kbr-scylla replied:
We'll have to fix cqlsh as a follow-up.

margdoc replied:
> Maybe it's because it uses WhiteListRoundRobinPolicy underneath?

Yes.
```
$ cqlsh <workdir>/cql.m

Traceback (most recent call last):
  File "/usr/local/bin/cqlsh", line 2357, in <module>
    main(*read_options(sys.argv[1:], os.environ))
  File "/usr/local/bin/cqlsh", line 2326, in main
    encoding=options.encoding)
  File "/usr/local/bin/cqlsh", line 463, in __init__
    load_balancing_policy=WhiteListRoundRobinPolicy([self.hostname]),
  File "/usr/local/lib/python2.7/dist-packages/cassandra/policies.py", line 426, in __init__
    for endpoint in socket.getaddrinfo(a, None, socket.AF_UNSPEC, socket.SOCK_STREAM)]
socket.gaierror: [Errno -2] Name or service not known
```

kbr-scylla reviewed (approved):
Given that:
- cqlsh doesn't work yet -- we'll need to fix it as a follow-up -- so there's no point adding it to the documentation now
- we don't have separate metrics for maintenance socket, instead, they are disabled
- there are no other outstanding comments

I am merging the PR.

avikivity commented:
> Given that:
> 
> * cqlsh doesn't work yet -- we'll need to fix it as a follow-up -- so there's no point adding it to the documentation now
> * we don't have separate metrics for maintenance socket, instead, they are disabled
> * there are no other outstanding comments
> 
> I am merging the PR.

We could remove the code that discriminates between them, it's not doing anything good.
